Add DocsVersionSwitcher

ashleemboyer opened the pull request:
## ✅ "Ready for Review" checklist

<!--
Check these items off in the GitHub PR UI as you complete them. If one of these items is not relevant to your PR, e.g. you do not have Asana ticket to go with them, you can remove them from the list.
-->

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## 🔗 Relevant links

<!--
Include links to the branch preview, Asana task, and Figma designs wherever possible to make reviewing your PR easier.

When including the preview link, make sure to remove any '.' characters from the branch name:
  - Example: ks.my-branch -> ksmy-branch
-->

- [Preview link](https://dev-portal-git-ambadd-docs-version-switcher-hashicorp.vercel.app/waypoint/docs) 🔎
- [Asana task](https://app.asana.com/0/0/1202150565840125/f) 🎟️

## 🗒️ What

<!--
Briefly list out the changes proposed in this PR.
-->

- Updates to `Disclosure`
  - adds `closeOnClickOutside` and `closeOnFocusOutside` props
  - adds conditional usage of the `useOnClickOutside` and `useOnFocusOutside` hooks
  - fixes a bug on the generated `containerClasses` ([explained in a PR annotation](https://github.com/hashicorp/dev-portal/pull/491/files#r884021474))
- Updates `NavigationDisclosure` to use the new `closeOnClickOutside` and `closeOnFocusOutside` `Disclosure` props
- Adds `DocsVersionSwitcher`
- Renders `DocsVersionSwitcher` in the bottom of the sidebar

## 🤷 Why

<!--
Describe why the changes proposed are needed. Some examples: new feature requested, refactor to make things easier later, styling tweaks requested by design, etc.
-->

- The `closeOnClickOutside` and `closeOnFocusOutside` props allow us to conditionally tell a `Disclosure` to close itself using the `useOnClickOutside` and `useOnFocusOutside` hooks. We need this to be conditional since not all `Disclosure`s should automatically close. For example, `AccordionDisclosure`s, should not automatically close themselves, but `NavigationDisclosure`s should.
- I made `DocsVersionSwitcher` responsible for rendering `null` rather than conditionally render the component where it's placed in `SidebarSidecar`. I think this makes the markup a little cleaner and delegates all the rendering logic to `DocsVersionSwitcher`. It also provided nice opportunities to provide `console` messages that will be helpful for debugging as we onboard more products to beta.

## 🛠️ How

<!--
Dive into the approach you took, list resources you referenced, detail other approaches you tried but didn't end up going with, etc.
-->

- `DocsVersionSwitcher` was pretty fun to build using `NavigationDisclosure`! It was nice not having to double-check and remember the markup needed, and the subcomponents made it really easy to style the different elements using our preferred practice of CSS modules and class names.
- Rendering `DocsVersionSwitcher` in the sidebar and making the versions data available to it required some changes in `src/layouts/sidebar-sidecar/server.ts`.
  - The `versions` array was already being fetched in this code, it just needed to be exposed to the layout via `layoutProps`.
  - There's a new `currentRootDocsPath` variable being declared, which represents the current `RootDocsPath` being viewed by the user. This is important for `DocsVersionSwitcher` to build a useful `aria-label` for the switcher, and it will also be useful for other areas in the app with this layout as well. For example: for us to base the breadcrumb text on the name of a root docs path: [[breadcrumb] naming inconsistencies](https://app.asana.com/0/1202110981600689/1202317183612789/f) 🎟️ 

## 📸 Design Screenshots

<!--
Include a screenshot or two and a link to the designs you referenced with this code. These are both helpful context for reviewers to understand what designs you were looking at when you were putting this code together.
-->

<details>
  <summary>Anatomy</summary>
  <a href="https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/?node-id=7864%3A84776">Figma link</a>
  <img width="1138" alt="CleanShot 2022-05-31 at 10 31 14@2x" src="https://user-images.githubusercontent.com/43934258/171198854-192a3479-992d-4574-9059-1fbb7df7f09c.png">
</details>

<details>
  <summary>Placement (zoomed-in)</summary>
  <a href="https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/?node-id=7819%3A78576">Figma link</a>
  <img width="1184" alt="CleanShot 2022-05-31 at 10 33 28@2x" src="https://user-images.githubusercontent.com/43934258/171199333-429d2e6e-0cad-49cd-91dc-fc907a5c1892.png">
</details>

<details>
  <summary>Placement (zoomed-out)</summary>
  <a href="https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/?node-id=7819%3A78733">Figma link</a>
  <img width="1030" alt="CleanShot 2022-05-31 at 10 34 57@2x" src="https://user-images.githubusercontent.com/43934258/171199618-6a45858f-5817-4196-8b99-6e63ceb55262.png">
</details>

## 🧪 Testing

<!--
Create a checklist for going through how to test your proposed changes. If there is anything to configure before interacting with the project in a browser, such as toggling feature flags, changing machine settings, or simulating behavior in browser dev tools, list those steps first.

- [ ] Step 1
- [ ] Step 2
- [ ] Step 3
- [ ] ...
-->

### General

- [ ] Go to [/waypoint/docs](https://dev-portal-git-ambadd-docs-version-switcher-hashicorp.vercel.app/waypoint/docs)
- [ ] You should see `DocsVersionSwitcher` in the bottom of the sidebar area
- [ ] Make sure the `DocsVersionSwitcher` stays in place when scrolling different areas of the page
- [ ] Make sure the `DocsVersionSwitcher` is located correctly at various viewport widths, including tablet & mobile
- [ ] Open the switcher
- [ ] Change the version
- [ ] The page URL should update
- [ ] The page content should show the correct content for the version selected
- [ ] The sidebar content should reflect the correct content for the version selected
- [ ] Sidebar URLs should have the currently selected version in them and go to the correct page
- [ ] Switch back to the latest version
- [ ] The url should update to no longer have the version in it

### 404

- [ ] Go to [/waypoint/docs/kubernetes/install](https://dev-portal-git-ambadd-docs-version-switcher-hashicorp.vercel.app/waypoint/docs/kubernetes/install)
- [ ] Switch to version `v0.1.x`
- [ ] You should see the version 404 page:
  <img width="791" alt="CleanShot 2022-05-31 at 10 46 34@2x" src="https://user-images.githubusercontent.com/43934258/171202299-a8259537-f9c2-4ad3-a8f7-e0145399665b.png">

## 💭 Anything else?

<!--
If there is anything you came across that you chose not to address in this PR but plan to soon, list those items here and any Asana tasks you created to go with them.
-->

- Worked deferred for now:
  - Showing the name of the current root docs path in the version switcher's label. This needs to be deferred for now because adding a tooltip to an interactive component is complex and needs to be done carefully. It will take some discussion on the engineering side before implementing this, so I've kept the text short for now.
- Work coming in follow-up PRs:
  - [Version alert at the top of the page](https://app.asana.com/0/1201147515801429/1202150565840128/f) 🎟️
  - [Update Sidebar Overview menu item to include version in href](https://app.asana.com/0/0/1202364251355937/f) 🎟️
  - [Include version in docs breadcrumb](https://app.asana.com/0/0/1202368498415342/f) 🎟️
    - Happy to resolve [[breadcrumb] naming inconsistencies](https://app.asana.com/0/1202110981600689/1202317183612789/f) 🎟️ during this work as well! 💖

ashleemboyer commented on src/components/disclosure/index.tsx:
```diff
@@ -96,7 +104,7 @@ const Disclosure = ({
 
     const handleRouteChangeStart = () => {
-      closeDisclosure
+      closeDisclosure()
```
🗒️ This was a bug! :lolsob:

ashleemboyer commented on src/components/disclosure/index.tsx:
```diff
@@ -108,13 +116,27 @@ const Disclosure = ({
+    closeOnFocusOutside && isOpen
+  )
+
+  // build the className prop to pass the `children` container
+  const containerClasses = classNames(
+    s.root,
+    typeof containerClassName === 'function'
+      ? containerClassName(isOpen)
+      : containerClassName
+  )
```
🗒️ This also had a bug. When I updated `containerClassName` to allow either a function or a string, I did not actually make the `else` branch append `containerClassName`. Only `s.root` was being applied!

ashleemboyer commented on src/layouts/sidebar-sidecar/sidebar-sidecar-layout.module.css:
```diff
@@ -82,6 +81,23 @@ for further details.
+  width: 100%;
+  z-index: 2;
+
+  @media (--dev-dot-tablet-down) {
+    /* manually calculated based on .sidebarWrapper's padding-bottom */
+    bottom: -18px;
+  }
```
📌 This doesn't seem super stable but I'm not sure what else to do for now.

zchsh replied:
💭 🎨   I think this causes issues on tablet-sized viewports:

![CleanShot 2022-05-31 at 12 03 30@2x](https://user-images.githubusercontent.com/4624598/171218876-e250a5e7-49a4-4284-be43-f83a7522e31e.png)

Personally I'd remove this, and leave `bottom: 0` across viewports.

I think if there are positioning issues with that in place, they're like a result of 🎟️ [[sidebar slash top nav] fix mismatch in CSS height of top bars vs reality, especially on mobile](https://app.asana.com/0/1202114367927919/1202316169578550/f)? Maybe we should make that ticket more of a priority - in the video attached, I mention just forcing the height on the elements in question to prevent these issues. Maybe that'd be a decent temporary fix? Let me know what you think, I'd be happy to make that change 👍

ashleemboyer replied:
This sounds good to me! I'll update and remove it.

I think the steps forward are:

- Remove the alpha header above the main navigation bar
- Update all items in the navigation header to have static units in CSS (no `rem`s)
  - This will make sure the height always stays the same
  - I don't love this in the long term, but it's a good bandaid until we find a long term solution
- Store the height value of the navigation header in a variable as was previously done

ashleemboyer replied:
Addressed in: https://github.com/hashicorp/dev-portal/pull/491/commits/41fc487233f3908972f3646ffe68b3e133f1d882

ashleemboyer commented on src/pages/_app.tsx:
```diff
@@ -18,7 +18,8 @@ import fetchLayoutProps from 'lib/_proxied-dot-io/fetch-layout-props'
 import { makeDevAnalyticsLogger } from 'lib/analytics'
 
-const showProductSwitcher = isPreview() && !isDeployPreview()
+// TODO - UPDATE BEFORE MERGING
+const showProductSwitcher = false
```
🚧 Have temporarily hidden this so it's easier to review the switcher

zchsh commented on src/components/docs-version-switcher/helpers.ts:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Extract the version from a path string, or an array of path segments
+ */
+function getVersionFromPath(path: string): string | undefined {
```
🎨  I think there's a version of this in `lib/get-version-from-path`.

I wonder if should `export const REGEX = /^v\d+\.\d+\.(\d+|\w+)$/i` from there as well.... or perhaps move this entire file to `lib/get-version-from-path`? (Maybe rename to `lib/version-path-helpers` or something? 🤔 )

ashleemboyer replied:
Oh nice! These are the exact same functions. I didn't realize that we had already pulled (at least) one of them over. 😊 I'll go ahead and lift up the others as well where needed. Great catch!

ashleemboyer replied:
Addressed in:

- https://github.com/hashicorp/dev-portal/pull/491/commits/1299686bbd2812e08b07f1138dd3bb8d1c1ddf38
- https://github.com/hashicorp/dev-portal/pull/491/commits/a738a55cea7d821e7098607ef4f5060501e500d0

BRKalow commented:
Really curious behavior, when using Chrome on iOS, the bottom action bar appears to only slide out of view when the document itself is scrolled, not other scroll containers. So when we open the mobile nav and scroll, the action bar isn't hidden and so it overlaps the version switcher.

![image](https://user-images.githubusercontent.com/1289701/171230609-d6759450-cafc-466c-97ec-f4ce755f51b2.png)

ashleemboyer commented on src/components/docs-version-switcher/index.tsx:
```diff
@@ -0,0 +1,121 @@
+
+  // Don't do anything if there aren't any options
+  if (!options || options.length === 0) {
+    console.warn('DocsVersionSwitcher has no `options` to render.')
```
Sounds good! And would love to see something like `debug` in the future. Updating now!

ashleemboyer replied:
Addressed in: https://github.com/hashicorp/dev-portal/pull/491/commits/c91bd820a6ba07130dfdbb9253486a486b4ff8ae

ashleemboyer commented:
@BRKalow that's so weird! I wonder if there are any hacks around that. 🤔

@zchsh I've addressed your feedback and this is now ready for re-review! Thank you for the suggestions 💖

BRKalow commented:
@ashleemboyer I think we still need to try and figure out the bottom spacing here. There's a big gap due to the `padding-bottom: 70px` on the sidebar wrapper component.
![Screen Shot 2022-06-01 at 9 58 46 AM](https://user-images.githubusercontent.com/1289701/171435879-0cadd130-8ff3-444b-a9da-f17c138671e2.png)

zchsh commented on src/layouts/sidebar-sidecar/sidebar-sidecar-layout.module.css:
```diff
@@ -82,6 +81,18 @@ for further details.
+  padding: 24px;
+}
+
+.docsVersionSwitcherWrapper {
```
Will +1 @BRKalow 's comment, realized [the earlier comment thread](https://github.com/hashicorp/dev-portal/pull/491#discussion_r885829405) wasn't totally accurate - it does seem the issues are happening because of a combination of the `70px` `padding-bottom` that Bryce mentioned, I think in addition to the issues with `var(--sticky-bars-height)` being inaccurate on smaller viewports due to text wrapping.

If we want to solve in this PR, I think a quick fix would be to set explicit heights on the top bars, to force sizes to reflect what we're setting in CSS (details in video on [this task](https://app.asana.com/0/1202114367927919/1202316169578550/f)).

However, I'm very okay with deferring getting the spacing right on mobile to a different PR - it feels tied up in a bunch of other tickets.

⛏️  To make sure this thread doesn't get lost, we could add a comment here in CSS in `sidebar-sidecar-layout.module.css` to note that mobile styles aren't quite done yet.
